Build changes and fixes for SnappyData 1.0

sumwale opened the pull request:
## Changes proposed in this pull request

- updated gradle to 3.5.1
- changed "provided" dependencies to compileOnly/testCompile as appropriate
- added copying of libgemfire*.so libraries to fix the UnsatisfiedLinkError
  that is always shown on server/lead startup; also copied to build outputs
  so IDEA runs will also be clean if "product" target has been run from command-line
  after a clean
- replaced fastutil by koloboke:
  - far leaner with its annotation based generated classes that are self-contained
    (18M for fastutil vs ~1.3M)
  - faster in benchmarks and more cacheline friendly structures (e.g. IntIntMap uses long array)
  - added couple of annotated interfaces IntObjectHashMap, ObjectLongHashMap
    with gradle-apt-plugin for cleaner compile-time support
  - updated NOTICE with the copyright information removing fastutil
- moved gemfire-joptsimple and gemfire-json into a new gemfire-util project
  that also includes the above koloboke classes
- corrected deprecation warnings from new gradle (<< replaced with doLast)
- updated gradle plugin versions to latest ones
- gradlew* script changes as from upstream release
  (as updated by ./gradlew wrapper --gradle-version 3.5.1)
- removed a check from NewLRUClockHand obsoleted by changes for SNAP-2012

## Patch testing

precheckin

## ReleaseNotes changes

NA

## Other PRs 

https://github.com/SnappyDataInc/snappydata/pull/846
https://github.com/SnappyDataInc/spark/pull/81

Commit:
Build changes and fixes for SnappyData 1.0

- updated gradle to 3.5.1
- changed "provided" dependencies to compileOnly/testCompile as appropriate
- added copying of libgemfire*.so libraries to fix the UnsatisfiedLinkError
  that is always shown on server/lead startup
- replaced fastutil by koloboke:
  - far leaner with its annotation based generated classes that are self-contained
    (18M for fastutil vs ~1.3M)
  - bit faster in benchmarks
  - more cacheline friendly structures (e.g. IntIntMap uses long array)
  - added couple of annotated interfaces IntObjectHashMap, ObjectLongHashMap
    with cleaner compile-time support using gradle-apt-plugin
  - updated NOTICE with the copyright information removing fastutil
- moved gemfire-joptsimple and gemfire-json into a new gemfire-util project
  that also includes the above koloboke classes
- corrected deprecation warnings from new gradle (<< replaced with doLast)
- updated gradle plugin versions to latest ones
- removed a check from NewLRUClockHand obsoleted by changes for SNAP-2012

Commit:
minor formatting change

Commit:
reducing forks for unit tests

Commit:
Merge remote-tracking branch 'origin/snappy/master' into buildFixesFor1.0

Commit:
Merge remote-tracking branch 'origin/snappy/master' into buildFixesFor1.0

Commit:
expand javadoc comment

Commit:
consolidated snappy properties in SystemProperties

Commit:
Merge remote-tracking branch 'origin/snappy/master' into buildFixesFor1.0

Commit:
updated dependency versions

minor update to *NativeCalls signature for jna update (generic arguments)

Commit:
move hadoop back to 2.7.3 to be fully compatible with upstream Spark

Commit:
fixing test failures

- add back morbay jetty and sun jersey packages to tests which are required
  by hbase/hdfs tests
- fix occasional failures in some other tests

Commit:
some minor test output changes

Commit:
split TransactionDUnit to reduce running time

- use common servers for tests in TransactionDUnit instead of starting/stopping
  in every test
- moved out tests that need special node settings (server-groups) into separate
  Transaction2DUnit; added extensions for RR and OffHeap like for TransactionDUnit

Together these two reduce running time from >25mins to ~3+2mins

- fixed table drop in JDBC.java to honour colocation chain
- dynamically update log-level in TestUtil method for cases where node is not being
  restarted at the end of test

Commit:
few more cleanups

Commit:
Merge remote-tracking branch 'origin/snappy/master' into buildFixesFor1.0

Commit:
split ClientServerDUnit for better balance in parallel test runs

Commit:
Merge remote-tracking branch 'origin/snappy/master' into buildFixesFor1.0

Commit:
setup HDFS store only once in EvictionByCriteriaTest

- updated EvictionByCriteriaTest to JUnit4 and use @BeforeClass, @AfterClass
  to create HDFS store only once for entire class
- also reduced the buckets in tables to 5 to reduce the time required to initialize
  table buckets and the corresponding parallel queue buckets

Above two changes reduce the running time of EvictionByCriteriaTest
and OffHeapEvictionByCriteriaTest from ~20mins to ~3mins

- split CreateHDFSStoreDUnit into two to better balance in parallel test runs

Commit:
[SNAP-2068] cleanup open transactions and TXContexts

- track hosted TXContexts for cleanup in cache close so that subsequent reconnects,
  if any, do not end up using stale TXContexts
- moved the separate snapshotTXState ThreadLocal inside TXContext to aid in cleanup
- changed code and test calls to use TXManagerImpl.getCurrentSnapshotTXState() and
  setSnapshotTXState() instead of direct ThreadLocal handle for above change
- added removal of TXContext from global list on thread close from
  ConnectionTable.releaseThreadsSockets() that is invoked by all GemFire thread pools
- added a ThreadFactory for SnappyThriftServer for above call which should be done for
  GemFire layer cleanup in any case; also added for legacy DRDAConnThread
- also added thread stats for thrift pool like done by other GemFire pools

Commit:
make naming of threads consistent

Commit:
Merge remote-tracking branch 'origin/SNAP-2068' into buildFixesFor1.0

Commit:
review comment: rollback TXState in thread close if cache is not closing

Commit:
correct initialization in JUnit4 tests

Commit:
Merge remote-tracking branch 'origin/SNAP-2068' into buildFixesFor1.0

Commit:
More store precheckin fixes

- added explicit rebalance call after server restarts in IndexPersistenceDUnit to avoid subsequent
  queries failing with bucketMoved exception (during iteration)
- removed stopAllVMs calls in IndexPersistenceDUnit else they stop the VMs without table cleanups
  and subsequent tests using lesser number of nodes hang waiting for the node to come up
- shutdown in EvictionByCriteriaTest first to close any lingering booted VM from previous tests

Commit:
Merge remote-tracking branch 'origin/snappy/master' into buildFixesFor1.0

Commit:
Merge remote-tracking branch 'origin/snappy/master' into buildFixesFor1.0

Commit:
Merge remote-tracking branch 'origin/snappy/master' into buildFixesFor1.0

Commit:
Merge remote-tracking branch 'origin/snappy/master' into buildFixesFor1.0

kneeraj reviewed (approved):
LGTM. Too big a change but of few types. So not very thorough review but got all the changes logic and intentions. So if precheckin is good it is all good.

sumwale commented:
@kneeraj Most of the large changes pushed subsequently after PR creation are to big tests being split into two (e.g. ClientServerDUnit), or tests being changed to boot up servers only once in the entire class to reduce the running time (e.g. TransactionDUnit). This became more prominent issue because the previous way of parallel dunit runs using internal method calls no longer worked with the new gradle, so switched to using the gradle's default way of parallel run.

However gradle's parallel run implementation has a weakness still not addressed (due to which the previous way was added) that it splits the tests into equal sized groups upfront instead of dynamically submitting to an executor pool. So then if some group contains big tests it will continue very long compared to other groups resulting in a long "tail" run where only one worker is running the tests for a while. This behaviour caused the running time of dunits to jump by nearly an hour and of junits by 15-20mins. So then changed the dunits/junits that were causing the long tail by splitting and/or using a single set of servers. No dunits are more by ~15mins while junits are about the same as before.

Commit:
Merge remote-tracking branch 'origin/snappy/master' into buildFixesFor1.0

Commit:
Updated conflation to deal with few cases

- use result of conflation for STOP with no START (e.g. CREATE ASYNC EVENTLISTENER with MANUALSTART as false)
- but still do conflation for DROPs even if there is no corresponding CREATE (e.g. for DROP IF EXISTS)
- add sizer for RegionValue in GfxdDDLRegion; delegate unknown types to calcMemSize in GfxdObjectSizer
  rather than failing
- minor updates to tests for better failure messages

Commit:
Use "isClosing()" for connection validity rather than stopped

Fixed a rare hang whereby a connection might be marked for closing but not
stopped, but a thread picks up the connection and is unable to receive
any reply since the connection has been de-registered.
